### PR TITLE
Evenly distribute exchange spooling data across different shards

### PR DESCRIPTION
**Pull request by linzebing**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

trino-exchange

> How would you describe this change to a non-technical end user or system administrator?

Sometimes we run into cases where we get throttled by S3. 

1. Supporting multiple buckets as exchange spooling destinations will distribute the requests to get around the throttling limits. Currently, data of different taskPartitions gets assigned to these buckets in a round-robin fashion.
2. Adding a random prefix to sink output path. Data output file path format: {randomizedPrefix}.{queryId}.{stageId}.{sinkPartitionId}/{attemptId}/{sourcePartitionId}_{splitId}.data

## Documentation

() No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java [hunk 1]
```diff
@@ -78,14 +78,14 @@
...
-            URI baseDirectory,
+            List<URI> baseDirectories,
             FileSystemExchangeStorage exchangeStorage,
             ExchangeContext exchangeContext,
             int outputPartitionCount,
             Optional<SecretKey> secretKey,
             ExecutorService executor)
     {
-        this.baseDirectory = requireNonNull(baseDirectory, "baseDirectory is null");
+        this.baseDirectories = requireNonNull(baseDirectories, "baseDirectories is null");
```
Comment (arhimondr) [inline plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:88]: Let's shuffle them here

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java [hunk 2]
```diff
@@ -242,8 +245,9 @@ private Multimap<Integer, FileStatus> getCommittedPartitions(URI committedAttemp
         }
     }
 
-    private URI getExchangeDirectory()
+    private URI getExchangeDirectory(int taskPartitionId)
     {
+        URI baseDirectory = baseDirectories.get(taskPartitionId % baseDirectories.size());
```
Comment (arhimondr) [inline plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:250]: Let's round-robin them

**Review by arhimondr — COMMENTED** [review]
LGTM % comments

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java (hunk 2 above)
Comment (linzebing) [reply plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:250]: Discussed offline, the current design requires a deterministic mapping from partitionId to bucket

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java [hunk 3]
```diff
@@ -78,14 +78,14 @@
...
-            URI baseDirectory,
+            List<URI> baseDirectories,
             FileSystemExchangeStorage exchangeStorage,
             ExchangeContext exchangeContext,
             int outputPartitionCount,
             Optional<SecretKey> secretKey,
             ExecutorService executor)
     {
-        this.baseDirectory = requireNonNull(baseDirectory, "baseDirectory is null");
+        this.baseDirectories = requireNonNull(ImmutableList.copyOf(baseDirectories), "baseDirectories is null");
```
Comment (arhimondr) [inline plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:88]: `ImmutableList.copyOf(requireNonNull(baseDirectories, "baseDirectories is null"))`

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeManager.java [hunk 4]
```diff
@@ -93,7 +89,8 @@ public Exchange createExchange(ExchangeContext context, int outputPartitionCount
                 throw new TrinoException(GENERIC_INTERNAL_ERROR, "Failed to generate new secret key: " + e.getMessage(), e);
             }
         }
-        FileSystemExchange exchange = new FileSystemExchange(baseDirectory, exchangeStorage, context, outputPartitionCount, secretKey, executor);
+        Collections.shuffle(baseDirectories);
```
Comment (arhimondr) [inline plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeManager.java:92]: This is not thread safe. Create a copy instead. Actually I would recommend creating a copy and shuffling inside the `FileSystemExchange` as it already creates a copy.

Comment (linzebing) [reply plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeManager.java:92]: Ah I forgot to update the PR.

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java [hunk 5]
```diff
@@ -78,29 +85,21 @@
...
-            URI baseDirectory,
+            List<URI> baseDirectories,
             FileSystemExchangeStorage exchangeStorage,
             ExchangeContext exchangeContext,
             int outputPartitionCount,
             Optional<SecretKey> secretKey,
             ExecutorService executor)
     {
-        this.baseDirectory = requireNonNull(baseDirectory, "baseDirectory is null");
+        this.baseDirectories = new ArrayList<>(ImmutableList.copyOf(requireNonNull(baseDirectories, "baseDirectories is null")));
```
Comment (arhimondr) [inline plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:95]: nit: shuffle first and then create an immutable copy

**Review by arhimondr — APPROVED** [review]

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java [hunk 6]
```diff
@@ -294,6 +303,8 @@ public ExchangeSourceStatistics getExchangeSourceStatistics(ExchangeSourceHandle
     @Override
     public void close()
     {
-        exchangeStorage.deleteRecursively(getExchangeDirectory());
+        for (int i = 0; i < baseDirectories.size(); ++i) {
+            exchangeStorage.deleteRecursively(getExchangeDirectory(i));
```
Comment (losipiuk) [inline plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java]: nit: passing `i` as `taskPartitionId` is somewhat ugly, and exploits the implementation details of `getExchangeDirectory` (that it uses `partitionId` modulo `number of base directories` to determine final directory shape).
A cleaner approach would be to have explicitly method `List<URI> getAllExchangeDirectories()` and use it in `close` and in `initialize`

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeModule.java [hunk 7]
```diff
@@ -37,8 +38,9 @@ protected void setup(Binder binder)
     {
         binder.bind(FileSystemExchangeManager.class).in(Scopes.SINGLETON);
 
-        FileSystemExchangeConfig fileSystemExchangeConfig = buildConfigObject(FileSystemExchangeConfig.class);
-        String scheme = URI.create(requireNonNull(fileSystemExchangeConfig.getBaseDirectory(), "baseDirectory is null")).getScheme();
+        List<URI> baseDirectories = buildConfigObject(FileSystemExchangeConfig.class).getBaseDirectories();
+        verify(baseDirectories.stream().map(URI::getScheme).distinct().count() == 1, "Base directories should have the same scheme");
```
Comment (losipiuk) [inline plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeModule.java:42]: `verify` is to check for assertions which are expected to never happen unless there is a bug in the code. Here we can legitly get multiple different schemas if user provides wrong configuration, right?
Please make it throw `TrinoException` with `CONFIGURATION_INVALID` error code and message which allows user (Trino cluster admin) to understand what is happenning.

File: core/trino-main/src/test/java/io/trino/operator/TestDeduplicatingDirectExchangeBuffer.java [hunk 8]
```diff
@@ -67,7 +67,7 @@ public void beforeClass()
         exchangeManagerRegistry = new ExchangeManagerRegistry(new ExchangeHandleResolver());
         exchangeManagerRegistry.addExchangeManagerFactory(new FileSystemExchangeManagerFactory());
         exchangeManagerRegistry.loadExchangeManager("filesystem", ImmutableMap.of(
-                "exchange.base-directory", System.getProperty("java.io.tmpdir") + "/trino-local-file-system-exchange-manager"));
```
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/operator/TestDeduplicatingDirectExchangeBuffer.java:70]: typo in commit message `Renanme`

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java [hunk 9]
```diff
@@ -250,10 +241,14 @@ private Multimap<Integer, FileStatus> getCommittedPartitions(URI committedAttemp
         }
     }
 
-    private URI getExchangeDirectory(int taskPartitionId)
+    private URI getTaskOutputDirectory(int taskPartitionId)
     {
         URI baseDirectory = baseDirectories.get(taskPartitionId % baseDirectories.size());
-        return baseDirectory.resolve(exchangeContext.getQueryId() + "." + exchangeContext.getExchangeId() + PATH_SEPARATOR);
+        String randomizedPrefix = randomizedPrefixes.computeIfAbsent(taskPartitionId, ignored -> UUID.randomUUID().toString().split("-")[0]);
```
Comment (losipiuk) [inline plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:247]: Using `Random.nextBytes` would be more straightforward than generating long UUID and cutting it down

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java (hunk 6 above)
Comment (linzebing) [reply plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java]: With the last commit, this part has been changed to delete all task output directories. I will create a follow up PR to delete the directories in batches.

File: plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java (hunk 9 above)
Comment (linzebing) [reply plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java:247]: `Random.nextBytes` will generate Strings that may violate S3's naming rules.

**Review by arhimondr — APPROVED** [review]

**Pull request merged**